client/rest-high-level/src/test/java/org/elasticsearch/client/CustomRestHighLevelClientTests.java

```diff
@@ -0,0 +1,236 @@
+                    .when(restClient)
+                    .performRequestAsync(eq(HttpGet.METHOD_NAME), eq(ENDPOINT), anyMapOf(String.class, String.class),
+                            any(HttpEntity.class), any(ResponseListener.class), anyVararg());
+        }
+    }
+
+    public void testCustomRequest() throws IOException {
+        final int length = randomIntBetween(1, 10);
+
+        CustomRequest customRequest = new CustomRequest();
+        customRequest.setValue(randomAlphaOfLength(length));
+
+        CustomResponse customResponse = execute(customRequest, restHighLevelClient::custom, restHighLevelClient::customAsync);
+        assertEquals(length, customResponse.getLength());
+        assertEquals(expectedStatus(length), customResponse.status());
+    }
+
+    private Void performRequestAsync(HttpEntity httpEntity, ResponseListener responseListener) {
+        try {
+            responseListener.onSuccess(performRequest(httpEntity));
+        } catch (IOException e) {
+            responseListener.onFailure(e);
+        }
+        return null;
+    }
+
+    private Response performRequest(HttpEntity httpEntity) throws IOException {
+        try (XContentParser parser = createParser(REQUEST_BODY_CONTENT_TYPE.xContent(), httpEntity.getContent())) {
+            CustomRequest request = CustomRequest.fromXContent(parser);
+
+            int length = request.getValue() != null ? request.getValue().length() : -1;
+            CustomResponse response = new CustomResponse(length);
+
+            ProtocolVersion protocol = new ProtocolVersion("HTTP", 1, 1);
+            RestStatus status = response.status();
+            HttpResponse httpResponse = new BasicHttpResponse(new BasicStatusLine(protocol, status.getStatus(), status.name()));
+
+            BytesRef bytesRef = XContentHelper.toXContent(response, XContentType.JSON, false).toBytesRef();
+            httpResponse.setEntity(new ByteArrayEntity(bytesRef.bytes, ContentType.APPLICATION_JSON));
+
+            RequestLine requestLine = new BasicRequestLine(HttpGet.METHOD_NAME, ENDPOINT, protocol);
+            return new Response(requestLine, new HttpHost("localhost", 9200), httpResponse);
+        }
+    }
+
+    /**
+     * A custom high level client that provides methods to execute a custom request and get its associate response back.
+     */
+    static class CustomRestClient extends RestHighLevelClient {
+
+        private CustomRestClient(RestClient restClient) {
+            super(restClient);
+        }
+
+        public CustomResponse custom(CustomRequest customRequest, Header... headers) throws IOException {
+            return performRequest(customRequest, this::toRequest, this::toResponse, emptySet(), headers);
+        }
+
+        public void customAsync(CustomRequest customRequest, ActionListener<CustomResponse> listener, Header... headers) {
+            performRequestAsync(customRequest, this::toRequest, this::toResponse, listener, emptySet(), headers);
```

Comment:
I think that what confuses me here is that we call performRequest and performRequestAsync here, why do we mock the rest client then? Wouldn't it be better to test that RestHighLevelClient subclasses can use performRequestAndParseEntity and performRequestAsyncAndParseEntity (which are private at the moment)